### ISSUE TITLE
coverity:  CID 73362 (#1 of 1): Untrusted value as argument (TAINTED_SCALAR)

### ISSUE THREAD

**Issue by adrianreber** [checkpoint-restore/criu#1224]
While trying to fix coverity warnings I was unsure if this is a false positive or not. Posting it here for further comments.

It seems most of the code has been written in 2012 by @cyrillos. Do you think this report is correct or can I mark it as invalid?

```
 49int do_pb_read_one(struct cr_img *img, void **pobj, int type, bool eof)
 50{
 51        char img_name_buf[PATH_MAX];
 52        u8 local[PB_PKOBJ_LOCAL_SIZE];
 53        void *buf = (void *)&local;
 54        u32 size;
 55        int ret;
 56
   1. Condition !cr_pb_descs[type].pb_desc, taking false branch.
 57        if (!cr_pb_descs[type].pb_desc) {
 58                pr_err("Wrong object requested %d on %s\n",
 59                        type, image_name(img, img_name_buf));
 60                return -1;
 61        }
 62
 63        *pobj = NULL;
 64
   2. Condition !!empty_image(img), taking false branch.
   3. Condition !!empty_image(img), taking false branch.
 65        if (unlikely(empty_image(img)))
 66                ret = 0;
 67        else
   4. tainted_data_argument: Calling function bread taints argument size. [show details]
 68                ret = bread(&img->_x, &size, sizeof(size));
   5. Condition ret == 0, taking false branch.
 69        if (ret == 0) {
 70                if (eof) {
 71                        return 0;
 72                } else {
 73                        pr_err("Unexpected EOF on %s\n",
 74                               image_name(img, img_name_buf));
 75                        return -1;
 76                }
   6. Condition ret < 4UL /* sizeof (size) */, taking false branch.
 77        } else if (ret < sizeof(size)) {
 78                pr_perror("Read %d bytes while %d expected on %s",
 79                          ret, (int)sizeof(size),
 80                          image_name(img, img_name_buf));
 81                return -1;
 82        }
 83
   7. Condition size > 1024UL /* sizeof (local) */, taking true branch.
   8. lower_bounds: Checking lower bounds of unsigned scalar size by taking the true branch of size > 1024UL.
 84        if (size > sizeof(local)) {
 85                ret = -1;
   CID 73362 (#1 of 1): Untrusted value as argument (TAINTED_SCALAR)9. tainted_data: Passing tainted variable size to a tainted sink.
 86                buf = xmalloc(size);
 87                if (!buf)
 88                        goto err;
 89        }
 90
```

**Comment by cyrillos** [comment]
I'll take a look today evening and will reply, ok?

**Comment by cyrillos** [comment]
This is false positive. I think we might use an explicit cast if it placate coverity.
```
[cyrill@grain criu] git diff
diff --git a/criu/protobuf.c b/criu/protobuf.c
index 206223ca5c1f..8f1c8f83a9d4 100644
--- a/criu/protobuf.c
+++ b/criu/protobuf.c
@@ -74,7 +74,7 @@ int do_pb_read_one(struct cr_img *img, void **pobj, int type, bool eof)
                               image_name(img, img_name_buf));
                        return -1;
                }
-       } else if (ret < sizeof(size)) {
+       } else if (ret < (int)sizeof(size)) {
                pr_perror("Read %d bytes while %d expected on %s",
                          ret, (int)sizeof(size),
                          image_name(img, img_name_buf));
```

**Comment by cyrillos** [comment]
Feel free to push up this diff.

**Comment by adrianreber** [comment]
Thanks. Well, if it is false positive it could be just marked in coverity as a false positive. But looking it the patch, it looks really simple so maybe it would help to make sure it does not appear in future coverity runs. I will probably submit a patch with a few additional scanning fixes in a few days.

**Comment by cyrillos** [comment]
Sure, thanks a lot!
